src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -0,0 +1,124 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace System.Diagnostics
+{
+    public sealed class ActivitySource : IDisposable
+    {
+        private static object s_syncObject = new object();
+        private static List<ActivitySource> s_activeSources = new List<ActivitySource>();
+        private ActivitySource() { throw new InvalidOperationException(); }
+
+        /// <summary>
+        /// Event to subscribe to get the notification when any ActivitySource object get created or released.
+        /// </summary>
+        public static event EventHandler<ActivitySourceEventArgs>? OperationEvent;
+
+        /// <summary>
+        /// Event to subscribe to get the notification when any Activity object get created or disposed.
+        /// </summary>
+        public event EventHandler<ActivitySourceEventArgs>? ActivityEvent;
+
+        /// <summary>
+        /// Construct an ActivitySource object with the input name
+        /// </summary>
+        /// <param name="name">The name of the ActivitySource object</param>
+        public ActivitySource(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            lock (s_syncObject)
+            {
+                Name = name;
+                s_activeSources.Add(this);
+                EventHandler<ActivitySourceEventArgs>? handlers = OperationEvent;
+                if (handlers != null)
+                {
+                    handlers(this, ActivitySourceEventArgs.s_sourceCreated);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the list of all created ActivitySource objects.
+        /// </summary>
+        public static IEnumerable<ActivitySource> ActiveList
+        {
+            get
+            {
+                lock (s_syncObject)
+                {
+                    foreach (ActivitySource source in s_activeSources)
+                    {
+                        yield return source;
```

Comment:
Yielding while holding a lock is generally very bad. The lock will end up being held around all of the consumer's work

Reply:
what you suggest to do instead? 

---
In reply to: [383014690](https://github.com/dotnet/runtime/pull/32660#discussion_r383014690) [](ancestors = 383014690)

Reply:
Don't yield while holding the lock ;-)  If you need to lock around getting the sources, they you need to either allow s_activeSources to be immutable such that you can snap off an immutable copy efficiently, or you need to make a copy while holding the lock and then return that after releasing the lock, or you need to exit the lock before yielding and re-enter after, which also means dealing with the fact that the list could change in between.

Reply:
Ok, I'll modify that to do copy and then yield outside the lock. Thanks @stephentoub for catching that.